Pull request for FLOE-412 (connection of data entry form and pie chart)

waharnum opened the pull request:
This supplies the model relaying between the data entry form and the pie chart component. Includes a basic demo where you can enter data and see the chart update in real time.

cindyli commented on demos/index.html:
```diff
@@ -14,19 +13,63 @@
+
 
         <title>Chart Authoring Tool</title>
+        <style>
```
Would be nice to create a src/css/style.css file for all css stylings.

cindyli commented on demos/index.html:
```diff
@@ -14,19 +13,63 @@
 
         <title>Chart Authoring Tool</title>
+        <style>
+            .floec-ca-pieChart-pie {
```
According to "Class name conventions" (http://docs.fluidproject.org/infusion/development/ClassNameConventions.html), "floec-" type of class names are used for programmatic manipulation of the DOM. Class names for styling usually start with "floe-".

cindyli commented on demos/index.html:
```diff
@@ -14,19 +13,63 @@
+                float: left;
+                width: 400px;
+            }
+            .floec-ca-pieChart-legend {
```
See the comment above regarding class name conventions. Same comment applies to line 47.

cindyli commented on demos/index.html:
```diff
@@ -14,19 +13,63 @@
+        </style>
     </head>
 
     <body>
```
There are 2 <body> opening tags in this file, see line 58.

cindyli commented on demos/index.html:
```diff
@@ -14,19 +13,63 @@
     <body>
+        <div id="floec-chartAuthoring">
+        <div id="floec-chartAuthoring-pieChart"></div>
+        <div id="floec-chartAuthoring-dataEntryPanel"></div>
```
Indentations are off at line 55, 56 as they are sub-elements of the div at line 54.

cindyli commented on src/js/chartAuthoring.js:
```diff
@@ -26,18 +26,56 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
                 type: "floe.chartAuthoring.dataEntryPanel",
-                createOnEvent: "onTemplatesLoaded",
-                container: "{chartAuthoring}.container",
+                createOnEvent: "onPieChartReady",
```
Is there a reason that the creation of the data entry panel is dependent on the creation of the pie chart? Shouldn't they be independently instantiated once their templates are ready?

waharnum replied:
I found that independently instantiating them caused this error related to the modelRelay configuration: 

> infusion-custom.js:12039 Uncaught Error: Assertion failure - check console for more details: Error in modelRelay record member "target": {floe.chartAuthoring.pieChart}.model.dataSet must be a reference to an existing component

I found this was also true if the modelRelay was configured in the pieChart component instead of the dataEntryPanel.

It's possible a different approach to construction could work for independent instantiation, but I couldn't see an obvious way to configure the modelRelay behaviour around this scenario.

cindyli replied:
Right, the model relay introduces the dependency. Thanks for the clarification.

cindyli commented on src/js/chartAuthoring.js:
```diff
@@ -26,18 +26,56 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
-                createOnEvent: "onTemplatesLoaded",
-                container: "{chartAuthoring}.container",
+                createOnEvent: "onPieChartReady",
+                container: "{dataEntryPanel}.container",
```
There should be a `selectors` block that defines selectors used by parts of this component. An example: https://github.com/fluid-project/chartAuthoring/blob/master/src/js/legend.js#L44-L51

cindyli replied:
Then, the container here can use something like ```container: "{chartAuthoring}.dom.dataEntryPanel". 

The same comment applies to the `pieChart` sub-component at line 55.

waharnum replied:
I knew there had to be a better approach to this then what I was using. Thanks!

cindyli commented on src/js/chartAuthoring.js:
```diff
@@ -26,18 +26,56 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+                        backward: "never"
+                    },
+                    events: {
+                        onCreate: "{chartAuthoring}.events.onPanelReady"
```
Would be nice to name space event listeners so they can be overridden when needed. Also, since there are other listeners defined in `floe.chartAuthoring.dataEntry` component, this listener should be fired with `priority: "last"`.

```
events: {
    "onCreate.escalate": {
        funcName: "{chartAuthoring}.events.onPanelReady.fire",
        priority: "last"
    }
}
```

cindyli commented on src/js/chartAuthoring.js:
```diff
@@ -26,18 +26,56 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+                        template: "{templateLoader}.resources.pieChart"
+                    },
+                    listeners: {
+                        "onPieChartReady.escalate": {
```
This listener can be simplified to:
`"onPieChartReady.escalate": "{chartAuthoring}.events.onPieChartReady"`

cindyli commented on src/js/chartAuthoring.js:
```diff
@@ -47,14 +85,43 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
-        }
+        },
+        {
+            source: "{that}.options.dataEntryPanel.container",
```
Once `selectors` block is in place, the distributeOptions from line 97-106 can be removed.

cindyli commented on src/js/pieChart.js:
```diff
@@ -68,21 +72,41 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+            legend: ".floec-ca-pieChart-legend"
+        },
+        listeners: {
+            "onCreate.renderTemplate": "floe.chartAuthoring.pieChart.renderTemplate"
```
It might make more sense to move this event firing to `floe.chartAuthoring.templateInjection` component since they are more related? In there, this line can be simplified to:

```
listeners: {
    "onCreate.injectTemplate": ...
    "onCreate.onTemplateInjected": {
        listener: "{that}.events.onTemplateInjected.fire",
        priority: "after:injectTemplate"
    }
}
```

Also, what about having a more specific event name such as `onTemplateRendered` or `onTemplateInjected` instead of `onParentReady`?

waharnum replied:
Both good ideas - other components making use of the templateInjection grade in the future may also require an event to indicate the template content has been injected.

cindyli commented on tests/js/chartAuthoringTests.js:
```diff
@@ -21,10 +29,14 @@ https://github.com/floe/universal/LICENSE.txt
         },
         components: {
             dataEntryPanel: {
+                container: ".floec-dataEntryPanel"
```
When comment https://github.com/fluid-project/chartAuthoring/pull/7#discussion_r40561201 is addressed, this line and line 35 can be removed.

cindyli commented on tests/js/chartAuthoringTests.js:
```diff
@@ -78,15 +126,25 @@ https://github.com/floe/universal/LICENSE.txt
         });
-        jqUnit.assertEquals("The dataEntryPanel has not been rendered", "", that.container.html());
+        jqUnit.assertEquals("The dataEntryPanel has not been rendered", null, that.dataEntryPanel);
+        jqUnit.assertEquals("The pieChart has not been rendered", null, that.pieChart);
```
Line 129, 130 are to verify sub-components are instantiated instead of being rendered. Can you adjust test messages. Also, we can use `jqUnit.assertNotUndefined` to verify defined variables.

waharnum commented:
This should be ready for another round of review or merge if there are no more issues to address (thanks @cindyli)

cindyli commented on demos/index.html:
```diff
@@ -14,19 +13,39 @@
-        <div id="floec-chartAuthoring"></div>
+        <div id="floec-chartAuthoring">
+            <div id="floec-chartAuthoring-pieChart"></div>
+            <div id="floec-chartAuthoring-dataEntryPanel"></div>
```
We could create a template for line 29, 30 as in the future it would be expanded with more elements such as editable title and description etc.

cindyli commented on demos/index.html:
```diff
@@ -14,19 +13,39 @@
 
         <script type="text/javascript">
             floe.chartAuthoring("#floec-chartAuthoring", {
+                selectors: {
```
The "selectors" block would more belong to the component defaults for `floe.chartAuthoring` as default options.

cindyli replied:
I see, you do have a "selectors" option defined in the defaults block of `floe.chartAuthoring`. We probably should match the html template selectors with selector names defined in there, to simplify our demo code by getting rid of this option.

cindyli commented on demos/index.html:
```diff
@@ -14,19 +13,39 @@
                     }
+                },
+                dataEntryPanel: {
+                    container: "#floec-chartAuthoring-dataEntryPanel"
```
The hard-coded id names can be replaced with `{chartAuthoring}.dom.dataEntryPanel`. This will help to restrict the container to use sub-elements of the parent container.

cindyli replied:
Also, this block belongs more to the defaults definition of `floe.chartAuthoring` component.

The same comment applies to line 48 for `pieChart` sub-component.

cindyli replied:
You also have correct containers defined for sub-components `dataEntryPanel` and `pieChart`. So can we get rid of line 44-49?

cindyli commented on src/js/templateInjection.js:
```diff
@@ -21,6 +21,10 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
                 "priority": "first"
+            },
+            "onCreate.templateInjected": {
+                "listener": "floe.chartAuthoring.templateInjection.templateInjected",
```
We can get rid of `floe.chartAuthoring.templateInjection.templateInjected()` by writing this line as,
`listener: "{that}.events.onTemplateInjected.fire"`

waharnum commented:
@cindyli I refactored as we discussed, so the chart authoring UI is templated. I also added some distributeOptions blocks, since the component tree is getting somewhat nested.

cindyli commented on demos/src/css/chartAuthoring-demo.css:
```diff
@@ -0,0 +1,25 @@
+.floe-ca-pieChart-pie {
+    float: left;
+    width: 400px;
```
In preparation of integrating UI Options into this demo, which is what we usually do for all our websites, "em" or "rem" instead of "px" should be used for HTML element sizes. So that elements are responsive to UIO's text size change.

https://wiki.fluidproject.org/display/fluid/How+to+make+your+page+more+responsive#Howtomakeyourpagemoreresponsive-Useemsfortextsize

Same to line 7.

waharnum replied:
At the moment this would only scale the container size - if we want the drawn pie to be responsive to em/rem text size changes, I can work on implementing https://issues.fluidproject.org/browse/FLOE-414 (I have a basic autoscale working that will set the pie size at create time based on the DIV container size)

cindyli replied:
Thanks to clarify. Yes, the scalable pie issue can be addressed in FLOE-414. Just feel it's a good pratice to always use "em" or "rem" for sizes.

waharnum replied:
Agree, implemented.

cindyli commented on src/js/chartAuthoring.js:
```diff
@@ -24,20 +28,75 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
         events: {
-            onTemplatesLoaded: null
+            onTemplatesLoaded: null,
+            ononChartAuthoringInterfaceReady: null,
```
Do you mean "onChartAuthoringInterfaceReady"? Duplicate "on".

cindyli commented on src/js/chartAuthoring.js:
```diff
@@ -24,20 +28,75 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+                                        func: "floe.chartAuthoring.dataEntriesToPieChartData"
+                                    },
+                                    forward: "liveOnly",
+                                    backward: "never"
```
Both `forward` and `backward` restrictions are unnecessary as,
1. The initial `{that}.model.dataEntries` is an empty object;
2.  `fluid.transforms.free` doesn't support inverted transformation.

cindyli commented on src/js/chartAuthoring.js:
```diff
@@ -24,20 +28,75 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+                        template: "{templateLoader}.resources.chartAuthoring"
+                    },
+                    listeners: {
+                        "onCreate.escalate": "{chartAuthoring}.events.ononChartAuthoringInterfaceReady.fire"
```
Shall `onTemplateInjected` be escalated instead of `onCreate`?

cindyli commented on tests/js/chartAuthoringTests.js:
```diff
@@ -78,15 +118,25 @@ https://github.com/floe/universal/LICENSE.txt
-                dataEntryPanel.options.resources[resourceName === "dataEntryPanel" ? "template": resourceName].resourceText);
-        });
-        jqUnit.assertNotEquals("The dataEntryPanel has been rendered", "", dataEntryPanel.container.html());
+    floe.tests.chartAuthoringTester.verifyTool = function (that) {
```
Can you also verify the new subcomponent `chartAuthoringInterface`, such as its instantiation and its template being passed correctly? Thanks.

waharnum replied:
Added.

cindyli commented:
@waharnum, can you update the license info in Gruntfile.js? It's still pointing to fluid-project/infusion repo. Thanks.

waharnum commented:
@cindyli done, license updated

cindyli commented:
Merged @ df106a8f9d353d9829d6d58b5d18b2e218bd24c9
